packages/react/src/components/PageHeader/PageHeader.tsx

```diff
@@ -489,14 +503,150 @@ const PageHeaderTabBar = React.forwardRef<
+  const tagSize = tags[0]?.size || 'md';
+
+  const tagsWithIds = useMemo(() => {
+    return tags.map((tag, index) => ({
+      ...tag,
+      id: tag.id || `tag-${index}`,
+    }));
+  }, [tags]);
+
+  const tagsContainerRef = useRef<HTMLDivElement>(null);
+  const offsetRef = useRef<HTMLDivElement>(null);
+  // To close popover when window resizes
+  useEffect(() => {
+    const handleResize = () => {
+      // Close the popover when window resizes to prevent unwanted opens
+      setOpenPopover(false);
+    };
+
+    window.addEventListener('resize', handleResize);
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, []);
+
+  // overflow items hook
+  const {
+    visibleItems = [],
+    hiddenItems = [],
+    itemRefHandler = () => {},
+  } = useOverflowItems<TagItem>(
+    tagsWithIds,
+    tagsContainerRef as React.RefObject<HTMLDivElement>,
+    offsetRef as React.RefObject<HTMLDivElement>
+  ) || {
+    visibleItems: [],
+    hiddenItems: [],
+    itemRefHandler: () => {},
+  };
+
+  const handleOverflowClick = useCallback((event: React.MouseEvent) => {
+    event.stopPropagation();
+    setOpenPopover((prev) => !prev);
+  }, []);
+
+  // Find TabList and TabPanels from children
+  let tabListElement: React.ReactNode = null;
+  let tabPanelsElement: React.ReactNode = null;
+
+  // Process children to extract TabList and TabPanels
+  React.Children.forEach(children, (child) => {
+    if (React.isValidElement(child)) {
+      const typedChild = child as React.ReactElement<{
+        children?: React.ReactNode;
+        displayName?: string;
+      }>;
+      const isPageHeaderTabs =
+        child.type === PageHeaderTabs ||
+        (typeof child.type === 'function' &&
+          (typedChild.type as any).displayName === 'PageHeader.Tabs') ||
+        (typeof child.type === 'function' &&
+          (typedChild.type as any).displayName === 'PageHeaderTabs');
+
+      if (isPageHeaderTabs) {
+        // Extract TabList and TabPanels
+        React.Children.forEach(typedChild.props.children, (tabChild) => {
+          if (React.isValidElement(tabChild)) {
+            const typedTabChild = tabChild as React.ReactElement<{
+              displayName?: string;
+            }>;
+
+            const tabChildType = typedTabChild.type;
+            const isTabList =
+              (typeof tabChildType === 'function' &&
+                (tabChildType as any).displayName === 'TabList') ||
+              (tabChildType as any)?.name === 'TabList';
+            const isTabPanels =
+              (typeof tabChildType === 'function' &&
+                (tabChildType as any).displayName === 'TabPanels') ||
+              (tabChildType as any)?.name === 'TabPanels';
+            if (isTabList) {
+              tabListElement = tabChild;
+            } else if (isTabPanels) {
+              tabPanelsElement = tabChild;
+            }
+          }
+        });
+      }
+    }
+  });
```

Comment:
Does this need to happen every render, even if they don't give us `tags`? Are there any other optimizations that could be done for when `tags` aren't provided?